src/tools/miri/cargo-miri/src/phases.rs

```diff
@@ -582,8 +582,10 @@ pub fn phase_rustdoc(mut args: impl Iterator<Item = String>) {
         .map_or(0, |verbose| verbose.parse().expect("verbosity flag must be an integer"));
 
     // phase_cargo_miri sets the RUSTDOC env var to ourselves, so we can't use that here;
-    // just default to a straight-forward invocation for now:
-    let mut cmd = Command::new("rustdoc");
+    // If available, use the one provided by bootstrap; otherwise, simply default to a straightforward
+    // invocation for now:
+    let rustdoc = env::var("RUSTDOC_REAL").unwrap_or("rustdoc".to_string());
+    let mut cmd = Command::new(rustdoc);
```

Comment:
Shouldn't we use the value of `RUSTDOC` as it was set in the original `cargo-miri` invocation?

Reply:
Yes that's why I didn't say to use rustdoc here, I said to use rustdoc from an earlier self-invocation of this binary. Something like this:
```diff
diff --git a/src/tools/miri/README.md b/src/tools/miri/README.md
index b043805291e..0f7187fe097 100644
--- a/src/tools/miri/README.md
+++ b/src/tools/miri/README.md
@@ -505,6 +505,8 @@ binaries, and as such worth documenting:
 * `MIRI_LOCAL_CRATES` is set by `cargo-miri` to tell the Miri driver which
   crates should be given special treatment in diagnostics, in addition to the
   crate currently being compiled.
+* `MIRI_RUSTDOC` is set and read by different phases of `cargo-miri` to remember the
+  value of `RUSTDOC` from before it was overwritten.
 * `MIRI_VERBOSE` when set to any value tells the various `cargo-miri` phases to
   perform verbose logging.
 * `MIRI_HOST_SYSROOT` is set by bootstrap to tell `cargo-miri` which sysroot to use for *host*
diff --git a/src/tools/miri/cargo-miri/src/phases.rs b/src/tools/miri/cargo-miri/src/phases.rs
index e547599d954..2521479a919 100644
--- a/src/tools/miri/cargo-miri/src/phases.rs
+++ b/src/tools/miri/cargo-miri/src/phases.rs
@@ -202,6 +202,9 @@ pub fn phase_cargo_miri(mut args: impl Iterator<Item = String>) {
     cmd.env("MIRI_BE_RUSTC", "target"); // we better remember to *unset* this in the other phases!
 
     // Set rustdoc to us as well, so we can run doctests.
+    if let Some(old_rustdoc) = env::var_os("RUSTDOC") {
+        cmd.env("MIRI_RUSTDOC", old_rustdoc);
+    }
     cmd.env("RUSTDOC", &cargo_miri_path);
 
     cmd.env("MIRI_LOCAL_CRATES", local_crates(&metadata));
@@ -581,9 +584,10 @@ pub fn phase_rustdoc(mut args: impl Iterator<Item = String>) {
     let verbose = std::env::var("MIRI_VERBOSE")
         .map_or(0, |verbose| verbose.parse().expect("verbosity flag must be an integer"));
 
-    // phase_cargo_miri sets the RUSTDOC env var to ourselves, so we can't use that here;
-    // just default to a straight-forward invocation for now:
-    let mut cmd = Command::new("rustdoc");
+    // phase_cargo_miri sets the RUSTDOC env var to ourselves, and puts a backup
+    // of the old value into MIRI_RUSTDOC. So that's what we have to invoke now.
+    let rustdoc = env::var("MIRI_RUSTDOC").unwrap_or("rustdoc".to_string());
+    let mut cmd = Command::new(rustdoc);
 
     let extern_flag = "--extern";
     let runtool_flag = "--runtool";
```

Reply:
Maybe call it `MIRI_ORIG_RUSTDOC` to make its purpose more clear.